Compiler flag to specify line ending

kmashint opened the pull request:
This is a first pass, unit tests still TODO, of adding a compiler flag to specify the line ending:
https://github.com/Microsoft/TypeScript/issues/1693
 --newLine NEWLINE                  Emit newline: 'CRLF' (dos) or 'LF' (unix).

I'll next add unit tests as noted in this follow-up issue but please note any comments on code thus far:
https://github.com/Microsoft/TypeScript/issues/2918

i've signed the contributor agreement.

msftclas commented:
Hi __@kmashint__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1984,6 +1984,18 @@
+        "category": "Message", 
+        "code": 6062
+    },
+    "Argument for --newLine option must be 'CRLF' or 'LF'.": {
```
'--newLine' in single quotes.

DanielRosenwasser commented on src/compiler/types.ts:
```diff
@@ -1681,6 +1682,12 @@ module ts {
     }
 
+    export const enum NewLineKind {
+        DEFAULT = 0,
```
Since each of the options are optional properties, I don't think we need a default.

CyrusNajmabadi replied:
Or, if we do, just call it 'Platform'.

DanielRosenwasser commented on src/compiler/program.ts:
```diff
@@ -91,6 +91,8 @@ module ts {
             }
         }
 
+        let newLine = [sys.newLine, "\r\n", "\n"][options.newLine ? Number(options.newLine) : 0];
```
This is kind of unclear at first sight. I might do 

```TypeScript
let newLine =
    options.newLine === NewLineKind.LF ? "\n" :
    options.newLine === NewLineKind.CRLF ? "\r\n" :
    sys.newLine;
```

CyrusNajmabadi replied:
+1

CyrusNajmabadi commented on src/compiler/commandLineParser.ts:
```diff
@@ -58,6 +58,13 @@ module ts {
         },
         {
+            name: "newLine",
+            type: { "crlf": NewLineKind.CRLF, "lf": NewLineKind.LF },
```
NewLineKind.LineFeed, .CarriageReturnLineFeed

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1984,6 +1984,18 @@
     },
+    "Emit newline: 'CRLF' (dos) or 'LF' (unix).": {
+        "category": "Message",
+        "code": 6061
```
Why not 6060?

kmashint replied:
Sure, I'll adjust.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1984,6 +1984,18 @@
         "category": "Error",
         "code": 6059
     },
+    "Emit newline: 'CRLF' (dos) or 'LF' (unix).": {
```
DOS, Unix

DanielRosenwasser commented on src/compiler/types.ts:
```diff
@@ -1650,6 +1650,7 @@ module ts {
         locale?: string;
         mapRoot?: string;
         module?: ModuleKind;
+        newLine?: string;
```
This should be a NewLineKind.

kmashint commented:
Adjusted for comments thus far.  I'll look into test cases tomorrow.

mhegazy commented on src/compiler/program.ts:
```diff
@@ -91,6 +91,11 @@ module ts {
             }
         }
 
+		let newLine =
```
formatting, and use const.

mhegazy commented:
Thanks @kamaljit for the change. We will need a test for the new flag. to add a test,
- add a new test file in test\cases\compiler\<testname>.ts with a comment at the top of `// @newline: LF` , another one with CRLF, and one for wrong input as well.
- add handling for the new switch in [harnes.ts](https://github.com/Microsoft/TypeScript/blob/master/src/harness/harness.ts#L953)
and in [fileMetadataNames](https://github.com/Microsoft/TypeScript/blob/master/src/harness/harness.ts#L1479)
- you would need to pass then the correct newline value to getNewLine in harness.ts
- jake runtests
- jake baseline-accept

kmashint commented:
I found and started to make the harness.ts adjustments; the runtests is taking forever so I'll check in the morning.

DanielRosenwasser commented:
> the runtests is taking forever

Yeah, there's been some perf drop in recent versions of Node that will make it...worse. I've been on 0.10.33. I've recently been working to parallelize the tests.

kmashint commented:
Stalled this week catching up on paid work ... should have some time on the weekend to work on unit tests.

kmashint commented:
I thought I had this working locally, but I'll try to find other examples of an expected failure in unit test cases.
1) compiler tests for tests/cases/compiler/newLineFlagWithCR.ts "before all" hook:
     Error: Unknown option for newLine: CR

DanielRosenwasser commented on src/harness/harness.ts:
```diff
@@ -822,7 +825,8 @@ module Harness {
             // the currentDirectory is needed for rwcRunner to passed in specified current directory to compiler host
-            currentDirectory?: string): ts.CompilerHost {
+            currentDirectory?: string,
+			newLineKind?: ts.NewLineKind): ts.CompilerHost {
```
Don't use tabs

DanielRosenwasser commented on tests/cases/compiler/newLineFlagWithCR.ts:
```diff
@@ -0,0 +1,4 @@
+// @newline: CR
```
We don't support CR so I guess we shouldn't have this test.

kmashint replied:
It was suggested to add a negative test for an invalid command-line arg but the test framework doesn't seem to support it so I'll remove the test.

DanielRosenwasser replied:
Sorry about any miscommunication - my understanding is that our harness doesn't support that sort of thing, though it's highly possible that @mhegazy might know something I'm not aware of.

DanielRosenwasser commented on src/harness/harness.ts:
```diff
@@ -1028,7 +1037,16 @@ module Harness {
-                            newLine = setting.value;
+                            if (setting.value.toLowerCase() === 'crlf') {
+                                options.newLine = ts.NewLineKind.CarriageReturnLineFeed;
+                            } else if (setting.value.toLowerCase() === 'lf') {
```
`else` on the next line

DanielRosenwasser commented on src/harness/harness.ts:
```diff
@@ -1028,7 +1037,16 @@ module Harness {
+                                options.newLine = ts.NewLineKind.CarriageReturnLineFeed;
+                            } else if (setting.value.toLowerCase() === 'lf') {
+                                options.newLine = ts.NewLineKind.LineFeed;
+                            } else if (setting.value === '\\n') {
```
This doesn't look right- none of the other options work this way.

DanielRosenwasser replied:
Oh I see what happened; any idea how many other tests use that flag?

kmashint replied:
A grep search revealed only the contextualTyping.ts uses this other @newline flag as in the comment, no matches for @newlines plural.
// Handle old usage, e.g. contextualTyping.ts:// @newline: \n

I could use newlineflag for this new option, or change the old seldom-used option to something more internal, e.g. @normalizeNewline since it seems to be used in normalizeLineEndings(...).

DanielRosenwasser replied:
I don't even know why the heck we ever introduced that - I think @danquirk might've added it after we futzed with baselining issues, but I think it would be safe to assume we can use the LF flag here.

kmashint commented:
I've introduced a @ normalizenewline option to work as the old one did in contextualTyping.js.  I'll adjust for merge conflicts tomorrow.

kmashint commented:
OK, I've rebased my changes on top of microsoft/TypeScript.  I tried using @ newline for both the new arguments and the old usage in contextualTyping.js but there were side-effects since the old usage affects not only the JS file but other debugging output files so I adjusted contextualTyping.js to use @ normalizenewline and all tests pass without further special treatment.

mhegazy commented:
Thanks @kmashint!

kmashint commented:
Happy to contribute in a small way, I hope others find it useful as well.
